### PR TITLE
docs: auto-generate contract ABI docs using `@b00ste/hardhat-dodoc` package

### PR DESCRIPTION
**Pull request by b00ste**
# What does this PR introduce?

## 📄 Documentation

- Use custom hardhat-dodoc package to parse `@custom:tags`, `constructor`, `receive` & `fallback` NatSpec comments.
- Create template for generating ABI docs .md files from the NatSpec of the contract.
- Format the NatSpec of the following contracts `LSP0ERC725Account.sol`, `LSP0ERC725AccountCore.sol` & `LSP14Ownable2Step.sol`. That was done with the purpose of testing the template for auto-generating ABI docs pages.

### PR Checklist

- [x] Wrote Tests
- [x] Wrote Documentation
- [x] Ran `npm run linter` (solhint)
- [x] Ran `npm run format` (prettier)
- [x] Ran `npm run build`
- [x] Ran `npm run test`

### CODE REVIEW

File: contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol [hunk 1]
```diff
@@ -93,11 +92,11 @@ abstract contract LSP0ERC725AccountCore is
     using Address for address;
 
     /**
-     * @dev Emits a `ValueReceived` event when receiving native tokens.
+     * @dev Executed:
```
Comment (CJ42) [inline contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol:95]: Suggested changes:

```suggestion
     * @dev The `receive()` function will run in the following scenarios:
     * - when the contract receives some native tokens without any additional data (`msg.data` is empty).
     * - when the contract receives an empty call without any value (`msg.data` and `msg.value` are both empty).
```

@YamenMerhi @skimaharvey can you confirm that's correct?

Comment (CJ42) [reply contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol:95]: Let's actually do the exact wordings and grammar in the Natspec comments in separate PRs for each contract.

File: contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol [hunk 2]
```diff
@@ -185,22 +184,21 @@ abstract contract LSP0ERC725AccountCore is
...
-     * on the owner accordinng to LSP20 - CallVerification specification
-     *
-     * Emits a {Executed} event, when a call is executed under `operationType` 0, 3 and 4
-     * Emits a {ContractCreated} event, when a contract is created under `operationType` 1 and 2
-     * Emits a {ValueReceived} event when receiving native tokens.
+     * - if the operation type is `STATICCALL` or `DELEGATECALL`, `value` SHOULD be 0.
+     * - `target` SHOULD be `address(0)` when deploying a contract.
+     * - MUST pass when called by the owner or by an authorised address that pass the verification check performed on the owner accordinng to [LSP20-CallVerification] specification
+     *
+     * @custom:events
```
Comment (CJ42) [inline contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol:198]: For the `@custom:events`, I would suggest that we do as follow to not have the verb _Emits_ repeated on each bullet points:

```
     * @custom:events Emitted Events:
     * - {Executed} event, when a call is executed under `operationType` 0, 3 and 4
     * - {ContractCreated} event, when a contract is created under `operationType` 1 and 2
     * - {ValueReceived} event when receiving native tokens.
```

Unless the `Emitted Events` term is added automatically in the template?

File: dodoc/config.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  if (textToFormat.length === 0) return "";
+  let formatedText: string = `**${title}**\n\n`;
+  if (textToFormat.startsWith("- ")) textToFormat = " " + textToFormat;
+  textToFormat.split(" - ").forEach((elem) => {
+    if (elem.trim().length !== 0) formatedText += `- ${elem.trim()}\n`;
+  });
+  return formatedText;
+};
+
+const genAdditionalInfo = (contract: string, code: string, type: string) => {
```
Comment (CJ42) [inline dodoc/config.ts:228]: Suggesting changing the function name

```suggestion
const generateAdditionalInfo = (contract: string, code: string, type: string) => {
```

File: dodoc/config.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+  code = code
+    .substring(0, code.indexOf(")") + 1)
+    .replace(`${type.toLowerCase()}`, "")
+    .trim();
+  if (!code.endsWith("()"))
```
Comment (CJ42) [inline dodoc/config.ts:233]: Suggesting adding a space + curly braces for the `if` statement

File: dodoc/config.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+      const middle = textToFormat.replace(start, "").replace(end, "");
+      formatedText = `${start}${middle
+        .split(",")
+        .map((elem) => `\n    ${elem.trim()}`)}\n${end}`;
+    } else {
+      const start = textToFormat.substring(0, textToFormat.indexOf(")") + 1);
+      const end = textToFormat.replace(start, "").trim();
+      formatedText = replaceAll(
+        `${start}${end.split(" ").map((elem) => {
+          if (elem.includes("(") || elem.includes(")")) return ` ${elem}`;
```
Comment (CJ42) [inline dodoc/config.ts:206]: it might read better here as a `if else` statement

File: package.json [hunk 6]
```diff
@@ -68,7 +68,7 @@
     "test:reentrancyinit": "hardhat test --no-compile tests/Reentrancy/ReentrancyInit.test.ts",
     "test:foundry": "forge test --no-match-test Skip -vvv --gas-report > gasreport.ansi",
     "clean": "hardhat clean && rm -rf module common",
-    "build": "hardhat compile",
+    "build": "hardhat dodoc && prettier -w docs",
```
Comment (CJ42) [inline package.json:71]: nice prettifying the markdown files 👌

**Review by YamenMerhi — APPROVED** [review]
Amazing! 👏 
Just one thing, when we view the the document it doesn't render the notes, is that normal ? 
Screen Shot 2023-06-08 at 10 52 26 AM

**Pull request merged**